src/Hamiltonians/Molecular.jl

```diff
@@ -0,0 +1,790 @@
+} <: AbstractOperatorColumn{A,T,O}
+    address::A # Contains address Psi_i, provided by operator_column
+    op::O   # Represent Hamiltonian itself, provided by operator_column
+    diag::T # < Psi_i | Psi_i >, calculated by diagonal_element
+    modes::M # Store the modes of address
+end
+
+function starting_address(h::MolecularHamiltonian)
+    return h.starting_address
+end
+
+function operator_column(
+    h::MolecularHamiltonian{T,A,D}, a::A
+)::MolecularHamiltonianOperatorColumn where {T<:Number,A<:FermiFS2C,D}
+    modes = full_mode_maps(a)
+
+    diag = zero(T)
+    diag_one_elec_int = one_electron_diagonal(h.fcidump.int1, modes.occupied)
+    diag_two_elec_int = two_electron_diagonal(h.fcidump.int2, modes.occupied)
+    diag = h.fcidump.int0 + diag_one_elec_int + diag_two_elec_int
+
+    return MolecularHamiltonianOperatorColumn{A,T,typeof(h),typeof(modes)}(
+        a, h, diag, modes
+    )
+end
+
+parent_operator(c::MolecularHamiltonianOperatorColumn) = c.op
+starting_address(c::MolecularHamiltonianOperatorColumn) = c.address
+
+function diagonal_element(
+    column::MolecularHamiltonianOperatorColumn{A,T,O,OD}
+) where {A<:FermiFS2C,T,O,OD}
+    return column.diag
+end
+
+function num_offdiagonals(column::MolecularHamiltonianOperatorColumn)
+    return length(offdiagonals(column))
+end
+
+function offdiagonals(column::MolecularHamiltonianOperatorColumn)
+    return MolecularHamiltonianOffDiagonals(column)
+end
+
+function random_offdiagonal(column::MolecularHamiltonianOperatorColumn)
+    ods = offdiagonals(column)
+    l = length(ods)
+    ri = rand(1:l)
+    state = linear_to_state(ods, ri)
+    r = iterate(ods, state)
+    if isnothing(r)
+        return (column.address, 0, 0)
+    end
+    return r[1][1], 1 / l, r[1][2]
+end
+
+"""
+    one_electron_diagonal(
+        int1::Matrix{<:Number},
+        occ_modes::NTuple{2,AbstractVector{FermiFSIndex}},
+    )::T where {T<:Number}
+
+Calculate the one body operator diagonal term ``⟨U|Ĥ₁|U⟩`` for a two-component Fermi Fock address.
+```math
+⟨U|Ĥ₁|U⟩ = ∑_{i,σ}^N ⟨U| h_{ii} a^†_{iσ} a_{iσ} |U⟩,
+```
+where the mode map of the two-component Fock address ``|U⟩`` is passed as `occ_modes`.
+"""
+function one_electron_diagonal(
+    int1::Matrix{T},
+    occ_modes::NTuple{2,AbstractVector{FermiFSIndex}},
+)::T where {T<:Number}
+    one_elec_int = zero(T)
+    for occ_mode in occ_modes
+        for i in occ_mode
+            one_elec_int += int1[i.mode, i.mode]
+        end
+    end
+    return one_elec_int
+end
+
+"""
+    two_electron_diagonal(
+        int2::Array{T,4},
+        occ_modes::NTuple{2,AbstractVector{FermiFSIndex}},
+    )::T where {T<:Number}
+
+Calculate the two body operator diagonal term ``⟨U|Ĥ₂|U⟩`` for a two-component Fermi Fock address.
+```math
+⟨U|Ĥ₂|U⟩ 
+    = ∑_{i < j,σᵢ,σⱼ}^{N} (V_{ij,ij} a^†_{i,σᵢ} a^†_{j,σⱼ} a_{j,σⱼ} a_{i,σᵢ} 
+    - V_{ij,ji} a^†_{i,σᵢ} a^†_{j,σⱼ} a_{i,σᵢ} a_{j,σⱼ} δ_{σᵢ,σⱼ})
```

Comment:
This formula is not correct as the LHS is a number and the RHS an operator. Also, I don't understand why the spins should be restricted to be the same in the second term. Should it not be the following?
```suggestion
⟨U|Ĥ₂|U⟩ 
    = ∑_{i < j,σ,τ}^{M} (V_{ij,ij} ⟨U| a^†_{i,σ} a^†_{j,τ} a_{j,τ} a_{i,σ} |U⟩ 
    - V_{ij,ji} ⟨U| a^†_{i,σ} a^†_{j,τ} a_{i,τ} a_{j,σ}|U⟩ )
```
NB: The sum go over `M` orbitals.

Reply:
The two-body operator term for occupied orbitals `i` and `j` is given by `J_ij - K_ij`, where `J` is the Coulomb integral and `K` is the exchange integral. The factor `δ_{σ,τ}` is required because when `i` and `j` belong to different spin channels, the exchange integral vanishes, which is reflected in the code.
```julia
sum_alpha_beta = zero(T)
for i in occ_modes[1]
    for j in occ_modes[2]
        sum_alpha_beta += int2[i.mode, j.mode, i.mode, j.mode]
    end
end
two_elec_int += sum_alpha_beta
```

Reply:
OK, thanks!